backend/src/main/java/gov/cdc/usds/simplereport/service/OktaService.java

```diff
@@ -0,0 +1,121 @@
+
+/**
+ * Handles all user/organization management in Okta
+ */
+@Service
+public class OktaService {
+
+    @Value("${okta.entity-mgmt.org-url}")
+    private String ORG_URL;
+    @Value("${OKTA_API_KEY}")
+    private String API_TOKEN;
+    @Value("${simple-report.authorization.role-prefix}")
+    private String ROLE_PREFIX;
+
+    Client _client;
+
+
+    public OktaService() {
+        _client = Clients.builder()
+            .setOrgUrl(ORG_URL) 
+            .setClientCredentials(new TokenClientCredentials(API_TOKEN))
+            .build();
+    }
+
+    public void createUser(ApiUser apiUser) {
+        Map<String,Object> userProfileMap = new HashMap<String,Object>();
+        userProfileMap.put("firstName",apiUser.getNameInfo().getFirstName());
+        userProfileMap.put("middleName",apiUser.getNameInfo().getMiddleName());
+        userProfileMap.put("lastName",apiUser.getNameInfo().getLastName());
+        // This is cheating. Suffix and honorific suffix aren't the same thing. Shhh.
+        userProfileMap.put("honorificSuffix",apiUser.getNameInfo().getSuffix());
+        // We assume login == email
+        userProfileMap.put("email",apiUser.getLoginEmail());
+        userProfileMap.put("login",apiUser.getLoginEmail());
+
+        UserBuilder.instance()
+                .setProfileProperties(userProfileMap)
+                .buildAndCreate(_client);
+    }
+
+    public void updateUser(String oldUsername, ApiUser apiUser) {
+        User user = _client.listUsers(oldUsername, null, null, null, null).single();
+        user.getProfile().setFirstName(apiUser.getNameInfo().getFirstName());
+        user.getProfile().setMiddleName(apiUser.getNameInfo().getMiddleName());
+        user.getProfile().setLastName(apiUser.getNameInfo().getLastName());
+        // Is it our fault we don't accommodate honorific suffix? Or Okta's fault they 
+        // don't have regular suffix? You decide.
+        user.getProfile().setHonorificSuffix(apiUser.getNameInfo().getSuffix());
+        // We assume login == email
+        user.getProfile().setEmail(apiUser.getLoginEmail());
+        user.getProfile().setLogin(apiUser.getLoginEmail());
+        user.update();
+    }
+
+    public void createOrganization(String externalId, String name) {
+        for (OrganizationRole role : OrganizationRole.values()) {
+            GroupBuilder.instance()
+            .setName(generateGroupName(externalId, role))
+            .setDescription(generateGroupDescription(name, role))
+            .buildAndCreate(_client);
+        }
+    }
+
+    public void addUserToOrganization(ApiUser apiUser, String organizationExternalId) {
+        User user = _client.listUsers(apiUser.getLoginEmail(), null, null, null, null).single();
+        String userId = user.getId();
+        // We assume that a user can only be a member of one organization at a time;
+        // if user is in any groups, remove that association first
```

Comment:
if that's the case, should i _not_ accept organization ID as an arg in the updateUser graphql mutation? seems it would be moot to include it if changing org raises an error?

Reply:
decided to remove orgId argument.